service/storage_service.cc

```diff
@@ -2658,20 +2658,207 @@ void storage_service::add_expire_time_if_found(inet_address endpoint, int64_t ex
+            std::vector<gms::inet_address> current_targets;
+            std::unordered_map<gms::inet_address, send_meta_data> metas;
+            size_t num_partitions_processed = 0;
+            size_t num_bytes_read = 0;
+            nr_sst_current += sst_processed.size();
+            auto reader = table.make_streaming_reader(s, full_partition_range, sst_set);
 
-    return distributed_loader::process_upload_dir(_db, _sys_dist_ks, _view_update_generator, ks_name, cf_name).finally([this, ks_name, cf_name] {
-        slogger.info("Done loading new SSTables for {}.{}", ks_name, cf_name);
+            try {
+                netw::messaging_service& ms = _messaging.local();
+                for (;;) {
+                    auto mf = reader(db::no_timeout).get0();
+                    if (mf) {
+                        bool is_partition_start = mf->is_partition_start();
+                        if (is_partition_start) {
+                            ++num_partitions_processed;
+                            auto& start = mf->as_partition_start();
+                            current_dk = make_lw_shared<dht::decorated_key>(start.key());
+
+                            current_targets = rs.get_natural_endpoints(current_dk->token());
+                            if (primary_replica_only && current_targets.size() > 1) {
+                                current_targets.resize(1);
+                            }
+                            slogger.trace("load_and_stream: ops_uuid={}, current_dk={}, current_targets={}", ops_uuid,
+                                    current_dk->token(), current_targets);
+                            for (auto& node : current_targets) {
+                                if (!metas.contains(node)) {
+                                    auto [sink, source] = ms.make_sink_and_source_for_stream_mutation_fragments(reader.schema()->version(),
+                                            ops_uuid, cf_id, estimated_partitions, reason, netw::messaging_service::msg_addr(node)).get0();
+                                    slogger.debug("load_and_stream: ops_uuid={}, make sink and source for node={}", ops_uuid, node);
+                                    metas.emplace(node, send_meta_data(node, std::move(sink), std::move(source)));
+                                }
+                            }
+                        }
+                        frozen_mutation_fragment fmf = freeze(*s, *mf);
+                        num_bytes_read += fmf.representation().size();
+                        parallel_for_each(current_targets, [&metas, &fmf, is_partition_start] (const gms::inet_address& node) {
+                            return metas.at(node).send(fmf, is_partition_start);
+                        }).get();
+                    } else {
+                        for (auto& [node, meta] : metas) {
+                            meta.finish();
+                        }
+                        break;
+                    }
```

Comment:
This should be out of the loop.

Reply:
It can be outside the loop. It is also natural to handle the !mf path directly. With while (auto mf = co_await reader(db::no_timeout)), it is better to do it outside the loop.

Reply:
> > Looks good. The streaming target will reshard if it gets partitions for the wrong shards?
> 
> We do not even reshard on the receiver. It uses the same code of streaming which uses the multi shard writer to send the parititon to correct shard.

That's what I meant.

> 
> > Please test with an unsharded sstable.
> 
> Can you explain more? What do you want to test?


Create an sstable with smp::count = 1, then load it into a cluster with smp::count > 1, verify it is split correctly.

Reply:
(I don't know why this reply got attached to this code chunk)

Reply:
OK.